### ISSUE TITLE
[8.0.x] Migration to karaf 4.4.1: java.lang.ClassNotFoundException: org.eclipse.jetty.websocket.server.WebSocketServerFactory

### ISSUE THREAD

**Issue by ida-greuelr** [ops4j/org.ops4j.pax.web#1753]
Hello pax-web developers,

we started migrating our existing software base from karaf 4.3.7 to karaf 4.4.1. Therefore we started using pax-web 8.0.6.
We have implemented a websocket enpoint according to the given example from karaf:
- https://github.com/apache/karaf/tree/main/examples/karaf-websocket-example

With version 4.3.7 and before everything is working straight forward.

Since the upgrade to 4.4.1 and pax-web 8.0.6 we are getting a ClassNotFoundException:
- Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.websocket.server.WebSocketServerFactory

during the httpService.registerServlet(...) call when karaf is starting and activating our Component.

Analogous to #1747  I have created and attached a simple but complete test project to simulate our setup. The provided README contains  step-by-step instructions to reproduce the error. 

Upon start of the karaf container, you will see an exception like this:
```log
websockettest  | 09:10:22.069 ERROR [features-3-thread-1] bundle com.mycompany.websockettest.suite.monitoring:1.0.0 (34)[com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS(24)] : The activate method has thrown an exception
websockettest  | java.lang.RuntimeException: javax.servlet.ServletException: java.lang.RuntimeException: Unable to load org.eclipse.jetty.websocket.server.WebSocketServerFactory
websockettest  | 	at org.eclipse.jetty.servlet.ServletHandler.initializeHolders(ServletHandler.java:804)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHandler.setServlets(ServletHandler.java:1556)
websockettest  | 	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.setServlets(PaxWebServletHandler.java:193)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHandler.addServlet(ServletHandler.java:932)
websockettest  | 	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.addServletWithMapping(PaxWebServletHandler.java:203)
websockettest  | 	at org.ops4j.pax.web.service.jetty.internal.JettyServerWrapper.lambda$visitServletModelChange$7(JettyServerWrapper.java:1212)
[...]
websockettest  | 	Suppressed: java.lang.Throwable
websockettest  | 		at org.ops4j.pax.web.service.spi.model.ServerModel.run(ServerModel.java:534)
websockettest  | 		at org.ops4j.pax.web.service.internal.HttpServiceEnabled.doRegisterServlet(HttpServiceEnabled.java:456)
websockettest  | 		at org.ops4j.pax.web.service.internal.HttpServiceEnabled.registerServlet(HttpServiceEnabled.java:353)
websockettest  | 		at org.ops4j.pax.web.service.internal.HttpServiceEnabled.registerServlet(HttpServiceEnabled.java:343)
websockettest  | 		at org.ops4j.pax.web.service.internal.HttpServiceProxy.registerServlet(HttpServiceProxy.java:120)
websockettest  | 		at com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS.activate(NotificationServiceWS.java:40)
[...]
websockettest  | Caused by: javax.servlet.ServletException: java.lang.RuntimeException: Unable to load org.eclipse.jetty.websocket.server.WebSocketServerFactory
websockettest  | 	at org.eclipse.jetty.websocket.servlet.WebSocketServlet.init(WebSocketServlet.java:147)
websockettest  | 	at javax.servlet.GenericServlet.init(GenericServlet.java:180)
websockettest  | 	at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.init(OsgiInitializedServlet.java:44)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHolder$Wrapper.init(ServletHolder.java:1345)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:632)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHolder.initialize(ServletHolder.java:415)
websockettest  | 	at org.eclipse.jetty.servlet.ServletHandler.initializeHolders(ServletHandler.java:799)
websockettest  | 	... 
[websockettest.zip](https://github.com/ops4j/org.ops4j.pax.web/files/9476765/websockettest.zip)
20 more
websockettest  | Caused by: java.lang.RuntimeException: Unable to load org.eclipse.jetty.websocket.server.WebSocketServerFactory
websockettest  | 	at org.eclipse.jetty.websocket.servlet.WebSocketServletFactory$Loader.load(WebSocketServletFactory.java:54)
websockettest  | 	at org.eclipse.jetty.websocket.servlet.WebSocketServlet.init(WebSocketServlet.java:140)
websockettest  | 	... 26 more
websockettest  | Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.websocket.server.WebSocketServerFactory
websockettest  | 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
websockettest  | 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
websockettest  | 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
websockettest  | 	at java.base/java.lang.Class.forName0(Native Method)
websockettest  | 	at java.base/java.lang.Class.forName(Class.java:467)
websockettest  | 	at org.eclipse.jetty.websocket.servlet.WebSocketServletFactory$Loader.load(WebSocketServletFactory.java:48)
websockettest  | 	... 27 more
```

**Comment by grgrzybek** [comment]
Mind that this example uses WebSocket servlet.
Pax Web 8 supports WebSockets using SCIs and I've added some examples with just annotations - without using `org.eclipse.jetty.websocket.servlet.WebSocketServlet`.

Especially because the solution should work with Tomcat and Undertow as well.

See WAB example here: https://github.com/ops4j/org.ops4j.pax.web/blob/web-8.0.7/samples/samples-war/war-websocket-jsr356
* it contains `@javax.websocket.server.ServerEndpoint` annotated class
* it doesn't use any special `WEB-INF/web.xml` at all

There's also an HttpService based example which registers websockets: https://github.com/ops4j/org.ops4j.pax.web/blob/web-8.0.7/pax-web-itest/pax-web-itest-server/src/test/java/org/ops4j/pax/web/itest/server/httpservice/WebContainerWebSocketsTest.java

**Comment by ida-greuelr** [comment]
Thanks for the fast response @grgrzybek. I will have a look at this asap. And will call back, if needed.

**Comment by grgrzybek** [comment]
Hmm, I've analyzed your example a bit more. And I think there's a place for improvements here.
While I've added many integration tests (HttpService, Whiteboard and WAB) to Pax Web 8 and I've spent **a lot** of time tweaking the usage of class loaders and TCCLs configurations, in your case TCCL is set to system class loader...

Your bundle starts and calls this in SCR `@Activate` (`com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS#activate()`:
```
...
	  at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2072)
	  at org.ops4j.pax.web.service.spi.model.ServerModel.run(ServerModel.java:550)
	  at org.ops4j.pax.web.service.internal.HttpServiceEnabled.doRegisterServlet(HttpServiceEnabled.java:456)
	  at org.ops4j.pax.web.service.internal.HttpServiceEnabled.registerServlet(HttpServiceEnabled.java:353)
	  at org.ops4j.pax.web.service.internal.HttpServiceEnabled.registerServlet(HttpServiceEnabled.java:343)
	  at org.ops4j.pax.web.service.internal.HttpServiceProxy.registerServlet(HttpServiceProxy.java:120)
	  at com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS.activate(NotificationServiceWS.java:40)
...
```

Pax Web has single configuration thread, which performs the initialization of the servlet:
```
...
	  at org.eclipse.jetty.websocket.servlet.WebSocketServletFactory$Loader.load(WebSocketServletFactory.java:47)
	  at org.eclipse.jetty.websocket.servlet.WebSocketServlet.init(WebSocketServlet.java:140)
	  at javax.servlet.GenericServlet.init(GenericServlet.java:180)
	  at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.init(OsgiInitializedServlet.java:44)
	  at org.eclipse.jetty.servlet.ServletHolder$Wrapper.init(ServletHolder.java:1345)
	  at org.eclipse.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:632)
	  at org.eclipse.jetty.servlet.ServletHolder.initialize(ServletHolder.java:415)
	  - locked <0x2514> (a org.ops4j.pax.web.service.jetty.internal.PaxWebServletHolder)
	  at org.eclipse.jetty.servlet.ServletHandler.initializeHolders(ServletHandler.java:799)
	  at org.eclipse.jetty.servlet.ServletHandler.setServlets(ServletHandler.java:1556)
	  - locked <0x2515> (a org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler)
	  at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.setServlets(PaxWebServletHandler.java:193)
	  at org.eclipse.jetty.servlet.ServletHandler.addServlet(ServletHandler.java:932)
	  at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.addServletWithMapping(PaxWebServletHandler.java:203)
	  at org.ops4j.pax.web.service.jetty.internal.JettyServerWrapper.lambda$visitServletModelChange$7(JettyServerWrapper.java:1212)
	  at org.ops4j.pax.web.service.jetty.internal.JettyServerWrapper$$Lambda$704/0x00000007c06445a0.accept(Unknown Source:-1)
	  at java.util.ArrayList.forEach(ArrayList.java:1511)
	  at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	  at org.ops4j.pax.web.service.jetty.internal.JettyServerWrapper.visitServletModelChange(JettyServerWrapper.java:1140)
	  at org.ops4j.pax.web.service.spi.task.ServletModelChange.accept(ServletModelChange.java:101)
	  at org.ops4j.pax.web.service.spi.task.Batch.accept(Batch.java:430)
	  at org.ops4j.pax.web.service.jetty.internal.JettyServerController.sendBatch(JettyServerController.java:168)
	  at org.ops4j.pax.web.service.internal.HttpServiceEnabled.lambda$doRegisterServlet$1(HttpServiceEnabled.java:521)
	  at org.ops4j.pax.web.service.internal.HttpServiceEnabled$$Lambda$698/0x00000007c069a000.run(Unknown Source:-1)
	  at org.ops4j.pax.web.service.spi.model.ServerModel.lambda$run$2(ServerModel.java:540)
	  at org.ops4j.pax.web.service.spi.model.ServerModel$$Lambda$623/0x00000007c062d7f0.get(Unknown Source:-1)
...
```

While there are many places in the above trace, where Pax Web can set TCCL, it doesn't do it - because I didn't see a reason for >2 years ;)

However, let me check if I can configure TCCL according to the element being registered.

**Comment by grgrzybek** [comment]
However, even if I added the TCCL (which I'll add to Pax Web code), your bundle misses `org.eclipse.jetty.websocket.server` import anyway. After adding it to `monitoring` bundle, everything seems fine:
```
karaf@root()> servlet-list 
Bundle ID │ Name                                                                              │ Class                                                                             │ Context Path(s) │ URLs           │ Type        │ Context Filter
──────────┼───────────────────────────────────────────────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────────┼─────────────────┼────────────────┼─────────────┼───────────────
117       │ cxf-osgi-transport-servlet                                                        │ org.apache.cxf.transport.servlet.CXFNonSpringServlet                              │ /               │ /cxf/*         │ HttpService │ -
157       │ com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceServlet │ com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceServlet │ /               │ /mywebsocket/* │ HttpService │ -
```

**Comment by grgrzybek** [comment]
BTW, servlets, which are started when the context itself starts (here: `org.apache.cxf.transport.servlet.CXFNonSpringServlet`) are initialized within correctly set TCCL.

**Comment by grgrzybek** [comment]
@ida-greuelr I can release 8.0.8 soon.

**Comment by ida-greuelr** [comment]
@grgrzybek Thank you for your ongoing effort. I also tried setting the right context class loader after [some google research](http://eriklievaart.com/web/osgiwebsockets.html) and started to have doubts about myself, because the refused to get solved. :see_no_evil: 
As long as our upgrade to karaf 4.4.1 is  blocked by the CXF issue #1747 it is not really urgent for us. Is there already a defined date for the release of 8.0.8 or will you do it based on demand?

**Comment by grgrzybek** [comment]
> Is there already a defined date for the release of 8.0.8 or will you do it based on demand?
I don't have a date set - but I want to release this week.

> I also tried setting the right context class loader
It wouldn't work inside `com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS#activate()`, because the actual registration is performed in another thread. It also wouldn't work in `com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceServlet#configure()`, because it's already after `org.eclipse.jetty.websocket.servlet.WebSocketServlet#init()` has been called. That's why the fix in 6ae2f239cdb09169ca2502e47681616eabb0df3c is needed.

**Comment by ida-greuelr** [comment]
> It wouldn't work inside `com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceWS#activate()`, because the actual registration is performed in another thread. It also wouldn't work in `com.mycompany.websockettest.suite.monitoring.websocket.NotificationServiceServlet#configure()`, because it's already after `org.eclipse.jetty.websocket.servlet.WebSocketServlet#init()` has been called. That's why the fix in [6ae2f23](https://github.com/ops4j/org.ops4j.pax.web/commit/6ae2f239cdb09169ca2502e47681616eabb0df3c) is needed.

Thank you for this comprehensive explanation.

For our development progress a release somewhere in the nearer future is good enough. We will have to wait for CXF 4.4.9 anyway :+1:

**Comment by ida-greuelr** [comment]
Feedback:
I updated our upgrade branch to use karaf 4.4.2 which has been released recently including pax-web 8.0.11.
Everything is working now as expected. @grgrzybek  **Thank you for your support and lightning-fast fix!!!**

**Comment by grgrzybek** [comment]
Thanks for your feedback @ida-greuelr ;)
